File: hearts-mike/src/Cards.hs
-- Das französische Blatt
module Cards where

import qualified Data.Set as Set
import Data.Set (Set)

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

data Suit = Diamonds | Clubs | Spades | Hearts
  deriving (Show, Eq, Ord)

-- |Liste aller Farben
allSuits :: [Suit]
allSuits = [Spades, Hearts, Diamonds, Clubs]

data Rank = Two | Three | Four | Five | Six | Seven | Eight | Nine | Ten
         | Jack | Queen | King | Ace
  deriving (Show, Eq, Ord)

-- |Liste aller Werte
allRanks :: [Rank]
allRanks = [Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, 
            Jack, Queen, King, Ace]

-- |Spielkarte
data Card = Card { suit :: Suit, rank :: Rank }
  deriving (Show, Eq, Ord)

-- |Eine Karte kann nur eine andere Karte gleicher Farbe nach Wert schlagen
cardBeats :: Card -> Card -> Maybe Bool
cardBeats c1 c2 = 
  if suit c1 == suit c2
  then Just (rank c1 > rank c2)
  else Nothing

cartesianProduct :: [a] -> [b] -> [(a, b)]
cartesianProduct list1 list2 =
  concatMap (\ el1 -> map (\ el2 -> (el1, el2)) list2) list1

-- |Das ganze Kartenspiel
deck :: [Card]
deck = map (uncurry Card) (cartesianProduct allSuits allRanks)

-- |Karten, die jemand auf der Hand hält
type Hand = Set Card

-- |Ist die Hand leer?
isHandEmpty :: Hand -> Bool
isHandEmpty hand = Set.null hand

-- |Enthält die Hand eine spezifische Karte?
containsCard :: Card -> Hand -> Bool
containsCard card hand = Set.member card hand

-- |Karte aus der Hand entfernen
removeCard :: Card -> Hand -> Hand
removeCard card hand = Set.delete card hand

-- |Leere Hand
emptyHand :: Hand
emptyHand = Set.empty

-- Spieler

data Player = Player { playerId :: String, playerName :: String }
  deriving Show

instance Eq Player where
  Player id1 _ == Player id2 _ = id1 == id2

instance Ord Player where
  compare (Player id1 _) (Player id2 _) = compare id1 id2

-- * Stich

-- Letzte Karte liegt oben
type Trick = [(Player, Card)]

-- leeren Stich herstellen
emptyTrick :: Trick
emptyTrick = []

-- ist Stich leer
trickEmpty :: Trick -> Bool
trickEmpty trick = null trick

-- wie groß ist der Stich?
trickSize :: Trick -> Int
trickSize trick = length trick

-- alle Karten desStich
cardsOfTrick :: Trick -> [Card]
cardsOfTrick trick = map snd trick

-- Karte auf den Stich legen
addToTrick :: Player -> Card -> Trick -> Trick
addToTrick player card trick = (player, card) : trick

-- die Karte des Stich, die bedient werden muß
leadingCardOfTrick :: Trick -> Card
leadingCardOfTrick trick = snd (last trick)

-- Hände der Spieler
type PlayerHands  = Map Player Hand

File: hearts-mike/src/main/scala/funar/hearts/Cards.scala
package funar.hearts

enum Suit {
  case Diamonds
  case Clubs
  case Spades
  case Hearts
}

object Suit {
  val all = Seq(Diamonds, Clubs, Spades, Hearts)
}

enum Rank {
  case Two
  case Three
  case Four
  case Five
  case Six
  case Seven
  case Eight
  case Nine
  case Ten
  case Jack
  case Queen
  case King
  case Ace
}

object Rank {
  given ordering: Ordering[Rank] = Ordering.by(_.ordinal)

  val all: Seq[Rank] = Seq(Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, 
                           Jack, Queen, King, Ace)
}

case class Card(suit: Suit, rank : Rank) {
  // Eine Karte kann nur eine andere Karte gleicher Farbe nach Wert schlagen
  def beats(other: Card): Option[Boolean] =
    if (this.suit == other.suit) {
      import Rank.ordering
      Some(Ordering[Rank].gt(this.rank, other.rank))
    } else
      None
}

object Card {

  // Seq: generalisierte Folgen // analog zu List<A> in Java
  // Seq: Interface
  // List: konkrete Implementierung 
  // vgl. List<A> Interface, Implementierung ArrayList<A>
  def cartesianProduct[A, B](as: Seq[A], bs: Seq[B]): Seq[(A, B)] =
    as.map { a => bs.map { b => (a, b) }}.flatten

  val deck : Seq[Card] = cartesianProduct(Suit.all, Rank.all).map { case (suit, rank) => Card(suit, rank) }
}

type Hand = Set[Card]

object Hand {
  val empty: Hand = Set.empty
}

// letzte Karte ist das erste Element
type Trick = List[(Player, Card)]

object Trick {
  val empty: Trick = List.empty

  val isEmpty(trick: Trick): Boolean = trick.isEmpty
  
  // def isEmpty(trick: Trick): Boolean = trick.isEmpty

  def cards(trick: Trick): Seq[Card] = trick.map(_._2) // 2. Komponente aus Tupel

  def add(trick: Trick, player: Player, card: Card) =
    (player, card) :: trick

  def leadingCard(trick: Trick): Card =
    trick.last._2

  def pretty(trick: Trick) = {
    def prettyOne(player: Player, card: Card): String = 
      prettyCard(card) ++ " from " ++ prettyPlayer(player)
    def p(s: Trick): String =
      s match {
        case Nil => ""
        case Seq((player, card)) => prettyOne(player, card)
        case ((player, card)::rest) => prettyOne(player, card) ++ ", " ++ p(rest)
      }
    p(trick.reverse)
  }

  def prettyPlayer(player: Player): String = player.name ++ "(" ++ player.id ++ ")"

  def prettyCard(card: Card): String =
    card.rank.toString ++ " of " ++ card.suit.toString

}


type PlayerHands = Map[Player, Hand]

